File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/utils/VehicleGravityAccelerationUtils.java
package com.github.aadvorak.artilleryonline.battle.utils;

import com.github.aadvorak.artilleryonline.battle.common.Position;
import com.github.aadvorak.artilleryonline.battle.model.RoomModel;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;

public class VehicleGravityAccelerationUtils {

    public static double getVehicleGravityAcceleration(VehicleModel vehicleModel, RoomModel roomModel) {
        return getLeftWheelAcceleration(vehicleModel, roomModel) + getRightWheelAcceleration(vehicleModel, roomModel);
    }

    private static double getRightWheelAcceleration(VehicleModel vehicleModel, RoomModel roomModel) {
        var wheelPosition = VehicleUtils.getRightWheelPosition(vehicleModel);
        var groundPosition = getWheelNearestGroundPositions(wheelPosition,
                vehicleModel.getSpecs().getWheelRadius(), roomModel);
        return getGravityAcceleration(wheelPosition, groundPosition, roomModel.getSpecs().getGravityAcceleration(),
                vehicleModel.getState().getAngle());
    }

    private static double getLeftWheelAcceleration(VehicleModel vehicleModel, RoomModel roomModel) {
        var wheelPosition = VehicleUtils.getLeftWheelPosition(vehicleModel);
        var groundPosition = getWheelNearestGroundPositions(wheelPosition,
                vehicleModel.getSpecs().getWheelRadius(), roomModel);
        return getGravityAcceleration(wheelPosition, groundPosition, roomModel.getSpecs().getGravityAcceleration(),
                vehicleModel.getState().getAngle());
    }

    private static double getGravityAcceleration(Position wheelPosition, Position groundPosition,
                                                 double gravityAcceleration, double vehicleAngle) {
        var groundAngle = Math.atan((groundPosition.getX() - wheelPosition.getX())
                / Math.abs(wheelPosition.getY() - groundPosition.getY()));
        return - 0.5 * gravityAcceleration * Math.sin(groundAngle) * Math.cos(vehicleAngle - groundAngle);
    }

    private static Position getWheelNearestGroundPositions(Position wheelPosition, double wheelRadius, RoomModel roomModel) {
        var groundIndexes = BattleUtils.getGroundIndexesBetween(wheelPosition.getX() - wheelRadius,
                wheelPosition.getX() + wheelRadius, roomModel);
        var i = groundIndexes.size() - 1;
        var groundPosition = BattleUtils.getGroundPosition(groundIndexes.get(i), roomModel);
        var groundDistance = groundPosition.distanceTo(wheelPosition);
        if (groundDistance <= wheelRadius) {
            return groundPosition;
        }
        for (i = groundIndexes.size() - 2; i >= 0; i--) {
            var position = BattleUtils.getGroundPosition(groundIndexes.get(i), roomModel);
            var distance = position.distanceTo(wheelPosition);
            if (distance <= wheelRadius) {
                return position;
            }
            if (distance <= groundDistance) {
                groundPosition = position;
                groundDistance = distance;
            }
        }
        return groundPosition;
    }
}

File: playground/src/utils/vehicle-gravity-acceleration-utils.ts
import { BattleUtils } from '@/utils/battle-utils'
import { VehicleUtils } from '@/utils/vehicle-utils'
import type { RoomModel, VehicleModel } from '@/data/model'
import type { Position } from '@/data/common'

export const VehicleGravityAccelerationUtils = {
  getVehicleGravityAcceleration(vehicleModel: VehicleModel, roomModel: RoomModel) {
    return (
      this.getLeftWheelAcceleration(vehicleModel, roomModel) +
      this.getRightWheelAcceleration(vehicleModel, roomModel)
    )
  },

  getRightWheelAcceleration(vehicleModel: VehicleModel, roomModel: RoomModel) {
    const wheelPosition = VehicleUtils.getRightWheelPosition(vehicleModel)
    const groundPosition = this.getWheelNearestGroundPositions(
      wheelPosition,
      vehicleModel.specs.wheelRadius,
      roomModel
    )
    return this.getGravityAcceleration(
      wheelPosition,
      groundPosition,
      roomModel.specs.gravityAcceleration,
      vehicleModel.state.angle
    )
  },

  getLeftWheelAcceleration(vehicleModel: VehicleModel, roomModel: RoomModel) {
    const wheelPosition = VehicleUtils.getLeftWheelPosition(vehicleModel)
    const groundPosition = this.getWheelNearestGroundPositions(
      wheelPosition,
      vehicleModel.specs.wheelRadius,
      roomModel
    )
    return this.getGravityAcceleration(
      wheelPosition,
      groundPosition,
      roomModel.specs.gravityAcceleration,
      vehicleModel.state.angle
    )
  },

  getGravityAcceleration(
    wheelPosition: Position,
    groundPosition: Position,
    gravityAcceleration: number,
    vehicleAngle: number
  ) {
    const groundAngle = Math.atan(
      (groundPosition.x - wheelPosition.x) / Math.abs(wheelPosition.y - groundPosition.y)
    )
    return -0.5 * gravityAcceleration * Math.sin(groundAngle) * Math.cos(vehicleAngle - groundAngle)
  },

  getWheelNearestGroundPositions(
    wheelPosition: Position,
    wheelRadius: number,
    roomModel: RoomModel
  ) {
    const groundIndexes = BattleUtils.getGroundIndexesBetween(
      wheelPosition.x - wheelRadius,
      wheelPosition.x + wheelRadius,
      roomModel
    )
    let i = groundIndexes.length - 1
    let groundPosition = BattleUtils.getGroundPosition(groundIndexes[i], roomModel)
    let groundDistance = BattleUtils.distance(groundPosition, wheelPosition)

    if (groundDistance <= wheelRadius) {
      return groundPosition
    }

    for (i = groundIndexes.length - 2; i >= 0; i--) {
      const position = BattleUtils.getGroundPosition(groundIndexes[i], roomModel)
      const distance = BattleUtils.distance(position, wheelPosition)
      if (distance <= wheelRadius) {
        return position
      }
      if (distance <= groundDistance) {
        groundPosition = position
        groundDistance = distance
      }
    }
    return groundPosition
  }
}
